### PR TITLE
Add runner for GDScript testing

### PR DESCRIPTION
**Pull request by vnen**
This is meant for testing the GDScript implementation, not for testing user scripts nor testing the engine using scripts.

Tests consists in a GDScript file and a .out file with the expected output. The .out file format is: expected status (based on the enum GDScriptTest::TestStatus) on the first line, followed by either an error message or the resulting output. Warnings are added after the first line, before the output (or compiler errors) if the parser pass without any error.

The test script must have a function called `test()` which takes no argument. Such function will be called by the test runner. The test should not have any dependency unless it's part of the test too. Global classes (using `class_name`) are registered before the runner starts, so those should work if needed.

Use the command `godot --gdscript-generate-tests godot-source/modules/gdscript/tests/scripts` to update the .out files
with the current output (make sure the output are the expected values before committing).

The tests themselves are part of the doctest suite so those can be executed with `godot --test`.

----

Thanks @Xrayez for the initial work porting the runner to doctest. This therefore supersedes #41074.

I moved the test suite to its own file because otherwise the tests would run twice.

### CODE REVIEW

File: tests/test_main.cpp [hunk 1]
```diff
@@ -131,6 +131,11 @@ int test_main(int argc, char *argv[]) {
 		memcpy(doctest_args[x], str, strlen(str) + 1);
 	}
 
+	test_context.setOption("order-by", "name");
+	test_context.setOption("abort-after", 20);
+	test_context.setOption("no-breaks", true);
+
+	// Allow the command-line to override doctest options above.
```
Comment (Xrayez) [inline tests/test_main.cpp:138]: This was removed in #43721.

Comment (vnen) [reply tests/test_main.cpp]: Ah, didn't realize I was changing `test_main.cpp`, probably my rebase re-added this.

**Comment by vnen** [comment]
Update to solve a couple of issues (separate commits since not directly related to this).

**Comment by vnen** [comment]
Another update to fix the CI. I didn't realize that it runs with a Mono build without the glue (we may want to revisit that if tests for Mono are introduced). Since I was initializing the languages to use GDScript, the absence of Mono glue made it crash. I changed to only initialize GDScript.

**Comment by akien-mga** [comment]
> I didn't realize that it runs with a Mono build without the glue (we may want to revisit that if tests for Mono are introduced).

On `3.x` it generates the glue, but for `master` it cannot be done in CI with a non-headless build yet (but #47414 might solve that).

**Comment by Xrayez** [comment]
I took #47157 and created a new test script for it:
```gdscript
# missing-constant-string-for-assert.gd

func test():
    var got = "GoDot"
    var expected = "Godot"
    assert(got == expected, "Error, expected: " + expected)
```
Then generated corresponding `out` files:
```
godot --gdscript-generate-tests modules/gdscript/tests/scripts
```
The resulting file looks like this (with the same name):
```
GDTEST_ANALYZER_ERROR
Expected constant string for assert error message.
```

Then run the GDScript test suite specifically:

```
godot --test --test-suite="*GDScript*" --success
```

```
modules\gdscript\tests\gdscript_test_runner.cpp(180): SUCCESS: CHECK( result.passed ) is correct!
  values: CHECK( true )
  logged: "D:/src/godot/gd4/modules/gdscript/tests/scripts/parser-errors/missing-contant-string-for-assert.gd"
          ""
```

So I'd say it's fairly functional!

Not sure about the requirement of specifying the path to GDScript test scripts for `--gdscript-generate-tests`, I think it should default to `"modules/gdscript/tests/scripts"` (just like a recent change to `--doctool`).

For testers: don't forget to compile the engine with `scons tests=yes`.

**Review by Xrayez — APPROVED** [review]

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**